Src/FluentAssertions/Equivalency/GenericDictionaryEquivalencyStep.cs

```diff
@@ -154,15 +154,81 @@ private static Type GetIDictionaryInterface(Type expectedType)
             return GetIDictionaryInterfaces(expectedType).Single();
         }
 
-        private static bool AssertSameLength<TSubjectKey, TSubjectValue, TExpectKey, TExpectedValue>(
+        private static bool AssertSameLength<TSubjectKey, TSubjectValue, TExpectedKey, TExpectedValue>(
             IDictionary<TSubjectKey, TSubjectValue> subject,
-            IDictionary<TExpectKey, TExpectedValue> expectation)
+            IDictionary<TExpectedKey, TExpectedValue> expectation) where TExpectedKey : TSubjectKey
         {
+            const string messageCore = "Expected {context:subject} to be a dictionary with {0} item(s), but found {1} item(s).";
+
+            if(expectation.Count != subject.Count)
+            {
+                // First, we gather missing and additional keys
+                List<TExpectedKey> missingKeys = new List<TExpectedKey>();
+                HashSet<TSubjectKey> presentKeys = new HashSet<TSubjectKey>();
+                
+                foreach (TExpectedKey expectationKey in expectation.Keys)
+                {
+                    if(subject.ContainsKey(expectationKey))
+                    {
+                        presentKeys.Add(expectationKey);
+                    }
+                    else
+                    {
+                        missingKeys.Add(expectationKey);
+                    }
+                }
+
+                IEnumerable<TSubjectKey> additionalKeys = subject.Keys.Except(presentKeys);
+
+                bool thereAreSomeMissingKeys = missingKeys.Count > 0;
+                bool thereAreSomeAdditionalKeys = additionalKeys.Any();
+
+                // Just missing keys
+                if(thereAreSomeMissingKeys && !thereAreSomeAdditionalKeys)
+                {
+                    return AssertionScope.Current
+                        .ForCondition(subject.Count == expectation.Count)
```

Comment:
You can leave out the `ForCondition`. 
The condition is always false is it was checked on line 163

Reply:
Thanks, I thought that it was still necessary, but you are right, it worked without ForCondition.